### ISSUE TITLE
Add MON$ATTACHMENT_ID column to the new MON$COMPILED_STATEMENTS table

### ISSUE THREAD

**Issue by hvlad** [FirebirdSQL/firebird#7534]
It is useful to know which attachment owns given compiled statement.
Even when shared metadata cache will be re-implemented in Super Server it still will be useful for Classic and Super Classic.

**Comment by asfernandes** [comment]
This will mean nothing in SuperServer with shared cache. No single attachment will own the statement.

Rather I see a need to have a monitoring table that show Firebird processes.

For example, a SuperServer with shared cache would be one process. It (not any of its attachment) would be the owner of a compiled statement.

**Comment by dyemanov** [comment]
Yes, we'll have to return NULL there for SS with shared cache. But this column will still be useful for CS/SC.

As for binding to processes, what about SuperClassic? It's a single (in the simplest case) process with multiple dbbs, thus having multiple caches.

**Comment by hvlad** [comment]
How do you going to explain users a lot of "duplicate" records in fb5 SS ? What is the use for it ?
I don't see NULL's  as a big problem in future shared cache SS, while in current SS I see it (problem) right now.

**Comment by asfernandes** [comment]
> How do you going to explain users a lot of "duplicate" records in fb5 SS ?

What duplicate records?

**Comment by hvlad** [comment]
Real sample, 120 connections, 2K+ procedures, 2k+ triggers

```
SELECT CS.MON$OBJECT_NAME, CS.MON$OBJECT_TYPE, COUNT( * )
  FROM MON$COMPILED_STATEMENTS CS
 WHERE CS.MON$OBJECT_NAME IS NOT NULL
GROUP BY 1, 2
HAVING COUNT(*) > 1
ORDER BY 3 DESC

```
MON$OBJECT_NAME    | MON$OBJECT_TYPE    | COUNT
--- | --- | ---
DATABASE_ON_CONNECT_1    | 2    | 89
DATABASE_ON_CONNECT_2    | 2    | 89
DATABASE_ON_TR_START    | 2    | 89
DATABASE_ON_TR_COMMIT_1    | 2    | 89
GL_EXCEPTION    | 5    | 89
GL_GDPR_JOG_OLVAS    | 5    | 89
TT_PROC_TRG_ROGZIT    | 5    | 89
TT_PROC_TRG_STOP    | 5    | 89
TT_PROC_TRG_START    | 5    | 89
CR_KONTAKT_T_BIUD1    | 2    | 21

**Comment by hvlad** [comment]
```
SELECT COUNT(*), T.RDB$TYPE_NAME
  FROM MON$COMPILED_STATEMENTS CS LEFT JOIN RDB$TYPES T 
    ON CS.MON$OBJECT_TYPE = T.RDB$TYPE 
   AND T.RDB$FIELD_NAME = 'RDB$OBJECT_TYPE'
GROUP BY 2
```

COUNT    | RDB$TYPE_NAME
--- | ---
146    |
3683   | PROCEDURE
8345   | TRIGGER
114    | UDF

**Comment by asfernandes** [comment]
> How do you going to explain users a lot of "duplicate" records in fb5 SS ?

Being honesty, saying that Firebird <= 5 SS do not do optimal memory management in regard to metadata/statements.

**Comment by hvlad** [comment]
Do you still object to add MON$ATTACHMENT_ID ? If not, I'm ready to do it.

**Comment by asfernandes** [comment]
I believe that addition of a column that is going to be deprecated or misused in future version should be must more well thought than it was, specially as there might be alternatives as I briefly explained.

**Comment by dyemanov** [comment]
I don't like it much, but I see no better choice so far.
